Index: src/org/mozilla/javascript/IRFactory.java
===================================================================
RCS file: /cvsroot/mozilla/js/rhino/src/org/mozilla/javascript/IRFactory.java,v
retrieving revision 1.122
diff -u -r1.122 IRFactory.java
--- src/org/mozilla/javascript/IRFactory.java	3 Mar 2009 16:02:09 -0000	1.122
+++ src/org/mozilla/javascript/IRFactory.java	8 May 2009 20:46:17 -0000
@@ -398,7 +398,7 @@
             // Bug: for code like "var obj={p:3};[obj.p]=[9];", "left" will
             // be ARRAYLITERAL with an embedded GETPROP. This causes errors
             // at codegen.
-            Kit.codeBug();
+            // Kit.codeBug();
             target = left;
         } else {
             target = transform(left);
Index: src/org/mozilla/javascript/Parser.java
===================================================================
RCS file: /cvsroot/mozilla/js/rhino/src/org/mozilla/javascript/Parser.java,v
retrieving revision 1.135
diff -u -r1.135 Parser.java
--- src/org/mozilla/javascript/Parser.java	7 Apr 2009 21:12:33 -0000	1.135
+++ src/org/mozilla/javascript/Parser.java	8 May 2009 20:46:19 -0000
@@ -3439,6 +3439,7 @@
               int type;
               if (nodeType == Token.GETPROP) {
                   type = Token.SETPROP;
+                  id.setType(Token.STRING);
               } else {
                   type = Token.SETELEM;
               }
